### ISSUE TITLE
RFE: add a session ID filter to the kernel's user filter

### ISSUE THREAD

**Issue by pcmoore** [linux-audit/audit-kernel#4]
There are times when admins need to restrict audit events based on the session information. The session ID is in the netlink credentials coming from user space. They just didn't get added to the user filter when they were added to netlink.  It would be nice to add them to the user filter.

**Comment by pcmoore** [comment]
While an RFE, this is a rather minor addition and a [feature page](https://github.com/linux-audit/audit-kernel/wiki/Template-RFE) isn't required, just the usual verbose patch description.  However, as with all new changes a corresponding patch for the [test suite](https://github.com/linux-audit/audit-testsuite) will be required for kernel patch acceptance.

**Comment by pcmoore** [comment]
@rgbriggs Considering the discussions we've had regarding the SESSIONID_SET flag, I think it might be a good idea to provide a RFE page in the wiki for this patch(set).  See the template page and the device auditing RFE page as an example; this should be quick work for this patchset.

* https://github.com/linux-audit/audit-kernel/wiki/Template-RFE
* https://github.com/linux-audit/audit-kernel/wiki/RFE-Device-Auditing

**Comment by rgbriggs** [comment]
RFE page ctreated:

- [(https://github.com/linux-audit/audit-kernel/wiki/RFE-Add-support-for-Session-ID-in-the-User-Filter)]

**Comment by pcmoore** [comment]
Thanks.  I did rename the wiki page to something a bit more concise:

* https://github.com/linux-audit/audit-kernel/wiki/RFE-Session-ID-User-Filter

I also don't see any mention of the SESSIONID_SET flag.

**Comment by rgbriggs** [comment]
A new patchset titled:
  Add support for session ID user filtering"
including:
  audit: add support for session ID user filter
  audit: add AUDIT_SESSIONID_SET support
  audit: add sessionid filter extension to feature bitmap
was posted:
  https://www.redhat.com/archives/linux-audit/2016-August/msg00038.html

**Comment by rgbriggs** [comment]
Userspace support patches were posted upstream.

A supporting prerequisite patchset for refactoring audit_get_feature() and adding support for loginuid_set from v3.10 will avoid merge conflicts:
https://www.redhat.com/archives/linux-audit/2016-August/msg00042.html

Also, support for
    https://github.com/linux-audit/audit-kernel/issues/5
posted
    https://www.redhat.com/archives/linux-audit/2016-August/msg00045.html
will address other potential merge conflicts.
    
The userspace support is in
    https://www.redhat.com/archives/linux-audit/2016-August/msg00048.html

The test case is
    https://github.com/rgbriggs/audit-testsuite/commit/c265e9fc5a7cd25315baec67f36d29ae0b11c020

**Comment by pcmoore** [comment]
@rgbriggs I think it might be nice to either add a new section to the RFE page providing auditctl examples on how to use it; this should help users/admins greatly.  Considering you have already done work on the tests this shouldn't be too difficult.

**Comment by pcmoore** [comment]
Quick update, I've merged a squashed version of @rgbriggs patches to the [audit-kernel/working-session_filter-v3](https://github.com/linux-audit/audit-kernel/tree/working-session_filter-v3) branch.

**Comment by pcmoore** [comment]
I'm also going to include the audit-kernel/working-session_filter-v3 branch in the [pcmoore/kernel-secnext](https://copr.fedorainfracloud.org/coprs/pcmoore/kernel-secnext/) builds from now until we merge this functionality into audit/next.  As soon as we hear from @stevegrubb that he is okay with the corresponding userspace patches I'll merge this into audit/next.

For reference, here is the pcmoore/kernel-secnext build with the patch included:
* https://copr.fedorainfracloud.org/coprs/pcmoore/kernel-secnext/build/442866

**Comment by rgbriggs** [comment]
I've added example usage to the RFE page.

**Comment by pcmoore** [comment]
Updated link to the RFE page: https://github.com/linux-audit/audit-kernel/wiki/RFE-Session-ID-User-Filter

**Comment by pcmoore** [comment]
Okay, my fault for not being specific; I think it would be good if we documented both example commands and their output on the RFE page.  For example, this is what we currently have:

    # date

... and this is what I think we should have:

    # date
    Tue Aug 30 15:07:48 EDT 2016

Make sense @rgbriggs ?

**Comment by rgbriggs** [comment]
@pcmoore Ok, output added to examples.

**Comment by rgbriggs** [comment]
Updated test at https://github.com/rgbriggs/audit-testsuite/tree/ghak4-test-for-sessionID-user-filter

**Comment by pcmoore** [comment]
At this point it would appear we are blocked waiting on @stevegrubb to review the associated userspace patches.

**Comment by pcmoore** [comment]
To summarize the current state of things ... We agreed on-list to drop the dedicated session ID set filter and use the (int)-1 sentinel value in its place.  It appears that we need to make some small changes to the existing patches to prevent this sentinel value from appearing in normal use, but I don't expect that too take long.

**Comment by pcmoore** [comment]
Further, @stevegrubb has merged the necessary userspace support code.

**Comment by pcmoore** [comment]
The -1 sentinel fix was merged via 833fc48d18ce3595990b405ae82a160b33028994.

**Comment by pcmoore** [comment]
I've tentatively merged the session ID filter support into the audit#next branch via 0489410368df46a8bfa31f9454c0158f79c6ff56.  If the tests and feature page are finished in time for the next merge window I'll include this patch in the pull request I send to Linus.

I'm going to keep this issue open until this is sent up to Linus.

**Comment by pcmoore** [comment]
I'm going to go ahead and close this out as the test component looks good, we are just waiting on the upcoming audit userspace v2.7 release, and the wiki page is in place and current.

 * https://github.com/linux-audit/audit-testsuite/pull/22
 * https://github.com/linux-audit/audit-kernel/wiki/RFE-Session-ID-User-Filter

**Comment by rgbriggs** [comment]
Posted a bug fix upstream for bug detected by code inspection implicating arch_f quick field pointer:
https://www.redhat.com/archives/linux-audit/2018-February/msg00024.html
https://lkml.org/lkml/2018/2/12/153
